src/Interpreters/executeQuery.cpp

```diff
@@ -591,65 +597,64 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
         {
             queue->push(ast, context);
 
-            BlockIO io;
             if (settings.wait_for_async_insert)
             {
                 auto timeout = settings.wait_for_async_insert_timeout.totalMilliseconds();
                 auto query_id = context->getCurrentQueryId();
                 auto source = std::make_shared<WaitForAsyncInsertSource>(query_id, timeout, *queue);
-                io.pipeline = QueryPipeline(Pipe(std::move(source)));
+                res.pipeline = QueryPipeline(Pipe(std::move(source)));
             }
 
-            return std::make_tuple(ast, std::move(io));
-        }
-
-        auto interpreter = InterpreterFactory::get(ast, context, SelectQueryOptions(stage).setInternal(internal));
-
-        std::shared_ptr<const EnabledQuota> quota;
-        if (!interpreter->ignoreQuota())
-        {
             quota = context->getQuota();
             if (quota)
+                quota->used(QuotaType::QUERY_INSERTS, 1);
```

Comment:
Should we also add these?
https://github.com/ClickHouse/ClickHouse/blob/0e2d0d70bae34dbaa04a44a0b901dacbb090f4c5/src/Interpreters/executeQuery.cpp#L629-L630

Reply:
Yes, it's worth to add it.